src/wrappers/themis/java/com/cossacklabs/themis/SecureCell.java

```diff
@@ -16,18 +16,529 @@
+         * You will have to provide the same context again during decryption.
+         * Usually this is some plaintext data associated with encrypted data,
+         * such as database row number, protocol message ID, etc.
+         *
+         * @param data      data to encrypt, must not be empty
+         * @param context   associated context, may be empty or {@code null}
+         *
+         * @return an object containing encrypted data and authentication token
+         * @throws InvalidArgumentException if data is empty
+         * @throws RuntimeException on internal encryption failure
+         *         (with {@link SecureCellException} as its cause)
+         */
+        @NotNull
+        @Contract(pure = true)
+        SecureCellData encrypt(byte[] data, @Nullable byte[] context);
+
+        /**
+         * Encrypts the provided message with associated context.
+         * <p>
+         * Data is encrypted and authentication token is produced separately.
+         * Use {@link #decrypt(byte[], byte[])} to decrypt the result later,
+         * providing both encrypted data and corresponding token.
+         * <p>
+         * This call is equivalent to {@link #encrypt(byte[], byte[])} with an empty associated context.
+         *
+         * @param data  data to encrypt, must not be empty
+         *
+         * @return an object containing encrypted data and authentication token
+         * @throws InvalidArgumentException if data is empty
+         * @throws RuntimeException on internal encryption failure
+         *         (with {@link SecureCellException} as its cause)
+         */
+        @NotNull
+        @Contract(pure = true)
+        SecureCellData encrypt(byte[] data);
+
+        /**
+         * Decrypts the provided message with associated context.
+         * <p>
+         * Secure Cell validates association with the context data, decrypts the message,
+         * and verifies its integrity using the provided authentication token.
+         * <p>
+         * You need to provide the same context data as provided to {@link #encrypt(byte[], byte[])}.
+         * You can also decrypt data encrypted with {@link #encrypt(byte[])} by using an empty context.
+         *
+         * @param data      encrypted data, cannot be empty
+         * @param token     authentication token, cannot be empty
+         * @param context   associated context, may be empty or {@code null}
+         *
+         * @return decrypted data if everything goes well
+         * @throws SecureCellException if decryption failed: for example,
+         *         if the data has been tampered with, if the secret is incorrect,
+         *         or if the associated context does not match the one used to encrypt the data
+         * @throws InvalidArgumentException if data or token is empty
+         */
+        @NotNull
+        @Contract(pure = true)
+        byte[] decrypt(byte[] data, byte[] token, @Nullable  byte[] context) throws SecureCellException;
```

Comment:
why not accept SecureCellData instead of two separate arguments with the same types? It adds new way to make mistake and pass args with incorrect order who not familiar with themis or other libraries and with common order of parameters? If we just add new alias for byte[] for token arg it will force users to think which of byte array they should convert to TokenType before decryption.

Reply:
> why not accept SecureCellData instead of two separate arguments with the same types?

First of all, to avoid constructing `SecureCellData` on each decryption. The users typically already have two separate `byte[]` arrays retrieved from the storage or network so it's not necessary to fuse them into `SecureCellData`.

Encryption returns `SecureCellData` because Java does not have multiple return values or tuples, we have no choice here. However, we'll make it so that it looks like a tuple return in Kotlin at least:

```kotlin
val encrypted, token = cell.encrypt(message)
```

(Not in this PR, it will be done later.)

And then, using `SecureCellData` does not make it much easier since its constructor *also* takes two `byte[]` arguments. If you don't know the order, there is no real difference:

```
byte[] decrypted = cell.decrypt(token, message)

byte[] decrypted = cell.decrypt(new SecureCellData(token, message))
```

(Both are incorrect, of course.)

However, Java devs typically use IDE which will provide contextual suggestions based on parameter names. Furthermore, the tuple return in Kotlin will also suggest the correct order. If everything fails, the developers will get an exception on the first attempt to decrypt data and then they'll learn the correct order.

So I think it's more readable to accept these arguments separately, without an intermediate `SecureCellData`.